rpc/core/pipe.go

```diff
@@ -47,7 +47,7 @@ type transport interface {
 type peers interface {
 	AddPersistentPeers([]string) error
 	DialPeersAsync([]string) error
-	NumPeers() (outbound, inbound, dialig int)
+	NumPeers() (outbound, inbound, dialing, unconditionalOutbound, unconditionalInbound int)
```

Comment:
is it really necessary to modify this function?

Reply:
you mean about `NumPeers()`? or only about interface?

The reason why I extend the `NumPeers()` function is to  check `MaxNumInboundPeers` with except `unconditionalInbound` count

```go
if !sw.IsUnconditionalPeer(p.NodeInfo().ID()) {
    // Ignore connection if we already have enough peers, except unconditional peer
    _, in, _, _, unconditionalIn := sw.NumPeers()
    if in-unconditionalIn >= sw.config.MaxNumInboundPeers {
        sw.Logger.Info(
            "Ignoring inbound connection: already have enough inbound peers",
```

and also modify `NumPeers` of interface `peers`  to fit the interface  `peers` to `Switch` struct

If there's anything wrong with me or missing, please let me know

Reply:
what if we just ignore unconditional peers in NumPeers?

in `NumPeers` func:

```go
			if !unconditional {
				outbound++
			}
```

Reply:
Yeah, that's what I thought at first,
But I think if we do that, the number of filtered out+in in `NetInfo()` and length of `p2pPeers.Peers().List()` is different, so `peers` list index out of range can occur. 

It could be resolved using  `p2pPeers.Peers().List()` not `out+in`  to make list `peers`
but there may be similar unforeseen circumstances.
What do you think?

```go
func NetInfo(ctx *rpctypes.Context) (*ctypes.ResultNetInfo, error) {
	out, in, _, _, _ := p2pPeers.NumPeers()
	peers := make([]ctypes.Peer, 0, out+in)
	for _, peer := range p2pPeers.Peers().List() {
		nodeInfo, ok := peer.NodeInfo().(p2p.DefaultNodeInfo)
		if !ok {
			return nil, fmt.Errorf("peer.NodeInfo() is not DefaultNodeInfo")
		}
		peers = append(peers, ctypes.Peer{
			NodeInfo:         nodeInfo,
			IsOutbound:       peer.IsOutbound(),
			ConnectionStatus: peer.Status(),
			RemoteIP:         peer.RemoteIP().String(),
		})
	}
```

Reply:
let's do that, yeah